pkg/apply/task/inv_set_task.go

```diff
@@ -54,79 +57,114 @@ func (i *InvSetTask) Identifiers() object.ObjMetadataSet {
-		klog.V(4).Infof("keep in inventory %d failed applies", len(applyFailures))
-		invObjs = invObjs.Union(applyFailures)
-
-		// If an object skipped apply and was previously stored in the inventory,
-		// then keep it in the inventory so it can be applied/pruned next time.
-		// It's likely that all the skipped applies are already in the inventory,
-		// because the apply filters all currently depend on cluster state,
-		// but we're doing the intersection anyway just to be sure.
-		applySkips := i.PrevInventory.Intersection(im.SkippedApplies())
-		klog.V(4).Infof("keep in inventory %d skipped applies", len(applySkips))
-		invObjs = invObjs.Union(applySkips)
-
-		// If an object failed to delete and was previously stored in the inventory,
-		// then keep it in the inventory so it can be applied/pruned next time.
-		// It's likely that all the delete failures are already in the inventory,
-		// because the set of resources to prune comes from the inventory,
-		// but we're doing the intersection anyway just to be sure.
-		pruneFailures := i.PrevInventory.Intersection(im.FailedDeletes())
-		klog.V(4).Infof("set inventory %d failed prunes", len(pruneFailures))
-		invObjs = invObjs.Union(pruneFailures)
-
-		// If an object skipped delete and was previously stored in the inventory,
-		// then keep it in the inventory so it can be applied/pruned next time.
-		// It's likely that all the skipped deletes are already in the inventory,
-		// because the set of resources to prune comes from the inventory,
-		// but we're doing the intersection anyway just to be sure.
-		pruneSkips := i.PrevInventory.Intersection(im.SkippedDeletes())
-		klog.V(4).Infof("keep in inventory %d skipped prunes", len(pruneSkips))
-		invObjs = invObjs.Union(pruneSkips)
-
-		// If an object is abandoned, then remove it from the inventory.
-		abandonedObjects := taskContext.AbandonedObjects()
-		klog.V(4).Infof("remove from inventory %d abandoned objects", len(abandonedObjects))
-		invObjs = invObjs.Diff(abandonedObjects)
-
-		// If an object is invalid and was previously stored in the inventory,
-		// then keep it in the inventory so it can be applied/pruned next time.
-		invalidObjects := i.PrevInventory.Intersection(taskContext.InvalidObjects())
-		klog.V(4).Infof("keep in inventory %d invalid objects", len(invalidObjects))
-		invObjs = invObjs.Union(invalidObjects)
-
-		klog.V(4).Infof("get the apply status for %d objects", len(invObjs))
-		objStatus := taskContext.InventoryManager().Inventory().Status.Objects
-
-		klog.V(4).Infof("set inventory %d total objects", len(invObjs))
-		err := i.InvClient.Replace(i.InvInfo, invObjs, objStatus, i.DryRun)
-
-		klog.V(2).Infof("inventory set task completing (name: %q)", i.Name())
+		if i.Destroy {
+			skippedDeletes := taskContext.InventoryManager().SkippedDeletes()
+			failedDeletes := taskContext.InventoryManager().FailedDeletes()
+			// if any deletes were failed/skipped, retain the inventory
```

Comment:
I suspect we should also keep the inventory if any reconciles failed.

This would slightly change the behavior of the ConfigSync finalizer, but the fact that CS ignores reconciler failures is a CS impl detail, not something we want to do in cli-utils or kpt.

In fact, because the CS finalizer is a preview annotation, we can still change how it works to wait for reconciles, even if the applier doesn't.

Reply:
i think you can also clean up the code here a little by extracting a `success() bool` so you can just do `if destroy && !success() { err = deleteInventory() } else { err = updateInventory() }; ch <- result(err); return`

Reply:
Done. Refactored the code and added reconcile failed/skipped check to the success condition